Add copy-to-clipboard buttons to offer templates

mnutt opened the pull request:
![sandstorm-clipboard](https://cloud.githubusercontent.com/assets/3705/12256045/2f416454-b8c8-11e5-8753-2cbeb07af608.gif)

Adds a {clipboardButton: 'left'} or {clipboardButton: 'right'} options to display the copy-to-clipboard button on one side or the other. It uses clipboard.js (inlined in offertemplate.html, let me know if you'd prefer a better way) which uses HTML5 clipboard api and falls back to highlighting the text for the user to hit "copy".

garply commented:
Hey admins! Can one of you let me know if it's safe to build this pull request? See https://alpha.sandstorm.io/grain/kB7bofATTL2jGuKSFwmqdA/GithubPRCommands for a list of commands.

paulproteus commented:
Garply, ok to test

paulproteus commented:
@zarvox probably wants to review this.

As a side note, https://github.com/sandstorm-io/sandstorm/wiki/Giving-credit-to-other-open-source-projects describes my current understanding of how we attribute other open source projects, so if this gets merged, we should follow those steps (or change the process).

paulproteus commented:
I am personally excited about seeing this clipboard functionality land in Sandstorm.

Additionally, we should presumably document it somewhere like https://docs.sandstorm.io/en/latest/developing/http-apis/#creating-an-offer-template - I'll let @zarvox provide feedback on the feature before asking for docs, but I wanted to bring that up at least.

kentonv commented:
This looks awesome. Agreed @zarvox should review (when he gets back).

zarvox commented on shell/public/clippy.svg:
```diff
@@ -0,0 +1,3 @@
+<svg height="16" width="14" xmlns="http://www.w3.org/2000/svg">
+  <path d="M2 12h4v1H2v-1z m5-6H2v1h5v-1z m2 3V7L6 10l3 3V11h5V9H9z m-4.5-1H2v1h2.5v-1zM2 11h2.5v-1H2v1z m9 1h1v2c-0.02 0.28-0.11 0.52-0.3 0.7s-0.42 0.28-0.7 0.3H1c-0.55 0-1-0.45-1-1V3c0-0.55 0.45-1 1-1h3C4 0.89 4.89 0 6 0s2 0.89 2 2h3c0.55 0 1 0.45 1 1v5h-1V5H1v9h10V12zM2 4h8c0-0.55-0.45-1-1-1h-1c-0.55 0-1-0.45-1-1s-0.45-1-1-1-1 0.45-1 1-0.45 1-1 1h-1c-0.55 0-1 0.45-1 1z" />
```
What's the source of this `clippy.svg`?  I'm not seeing it in https://github.com/zenorocha/clipboard.js

zarvox commented on shell/public/offer-template.html:
```diff
@@ -10,20 +10,89 @@
+    <div id="clipboard" data-clipboard-target="#text"></div>
+    <div id="clipboardSuccess">copied</div>
+<script>
+!function(t){if("object"==typeof exports&&"undefined"!=typeof module)module.exports=t();else if("function"==typeof define&&define.amd)define([],t);else{var e;e="undefined"!=typeof window?window:"undefined"!=typeof global?global:"undefined"!=typeof self?self:this,e.Clipboard=t()}}(function(){var t,e,n;return function t(e,n,r){function o(a,c){if(!n[a]){if(!e[a]){var s="function"==typeof require&&re ...
```
I kinda wish there were a better way to do this than inlining the minified `clipboard.js`.  I guess there's not a great solution short of adding some sort of build step to include, which seems like a big ask.

For now, could you add a comment with a link to the source here, so it's clearer what this is and where it came from?

We'll also need to add appropriate attribution for `clipboard.js` and `clippy.svg` in `LICENSE` - could you do that following the general pattern in that file?

zarvox commented on shell/public/offer-template.html:
```diff
@@ -10,20 +10,89 @@
+    <div id="container">
+      <pre id="text"></pre>
+    </div>
+    <div id="clipboard" data-clipboard-target="#text"></div>
```
I'm not sure how accessible this `<div>` is - can it be made into a `<button>` with an accessible label like "Copy" so it'll appear in the `taborder` and so folks like Nolan and other keyboard users will be able to take advantage of this?

zarvox commented:
Big :+1: on the idea.  Thanks for building this!

I'd appreciate a couple additions for license reasons, and if you can make the button more accessible, that would be fantastic. :)

mnutt commented:
Ok, I've updated the LICENSE and switched to a `<button>`, although I realized I'm actually not quite sure how to get it into the `taborder`, especially since it's in an iframe. @zarvox do you know what attribute I'd add?

Also, I moved clipboard.js to an external file. If we started accumulating more js files we'd probably need some sort of build/concat step, but as it is I think it's ok?

zarvox commented on shell/public/offer-template.html:
```diff
@@ -10,20 +10,92 @@
+    <div id="container">
+      <pre id="text"></pre>
+    </div>
+    <button id="clipboard" label="Copy" data-clipboard-target="#text"></button>
```
I could be mistaken, but I think you want `aria-label` here, not `label`.  I couldn't find an attribute `label` anywhere in the HTML5 spec.

zarvox commented:
The iframe itself is in the taborder, and once you tab into the iframe, you'll traverse all of the focusable elements within.  Since `<button>` is added to the taborder by default, I think this works fine.

I'm okay with making a blocking request for script on the first fetch of `offer-template.html`, especially since if the fetch fails it gracefully degrades to the old template's behavior.

So I think this just needs s/label/aria-label/ over by my last inline comment and we're good to go.

mnutt commented:
Alright, updated. I'm happy to take a stab at the documentation this weekend.

kentonv commented:
Garply, retest this please

mnutt commented:
@paulproteus I wrote a bit of documentation, let me know if you think there is anything else to add.

paulproteus commented:
The only thing I'd add to these docs is an `img src` that shows the clipboard button. That is not a show-stopper.

Honestly, the animated GIF you created in this pull request would be great as part of the docs, too.

Neither is a show-stopper. If you want, you can consider adding the animated GIF to the bottom or the top of this section: "Creating an offer template"

In fact, since I am currently putting images for the docs in a separate Davros grain, and using static publishing to get them published, here, have a URL you can use:

https://alpha-evgl4wnivwih0k6mzxt3.sandstorm.io/offer-template-video-by-mnutt.gif

and you can add it to the docs if you want, and if not, that's OK too.

If you do add it to the docs in here, I will take that as you licensing under the same terms as your other contributions to Sandstorm.

paulproteus commented:
+1 from me for docs/ changes

paulproteus commented:
@zarvox I guess we await a refreshed +1 from you.

zarvox commented:
:+1: from me.  Thanks again!

paulproteus commented:
Great - @kentonv I believe this is ready for your final review/merge.

kentonv commented:
Ugh apparently somehow this dropped off my list after I told Garply to retest.

So sorry for another delay, but @zarvox ran an auto-formatter across the codebase and now all our PRs have merge conflicts.

BTW I wonder how @neynah feels about the clipboard icon or if she wants to make a replacement.

Also which browsers has this been tested in? As I understand it different browsers have different rules about when you're allowed to invoke a "copy" command from script.

neynah commented:
Sorry for the delay, I missed this! I'd like to simplify/redesign the icon & I also think it should consistently be justified right.

I'm OK with using this icon for now until I design a new one this week.

zarvox commented:
@kentonv per https://clipboardjs.com/ this works in Chrome 42+, Firefox 41+, IE 9+, Opera 29+, and degrades to using the Selection API to select the contents to be copied otherwise.  I think that's perfectly fine, and is an improvement over the status quo, which has neither selection nor clipboard support.

FWIW, @mnutt I'm happy to rebase this atop master, since as far as I'm concerned it's ready to merge and I'm responsible for breaking everything - if you haven't replied or rebased this branch yourself by Friday, I'll just do it as a blur day task.

Thanks again!

kentonv commented:
@zarvox Be sure to use Nena's clipboard icon.

garply commented:
Hey admins! Can one of you let me know if it's safe to build this pull request? See https://alpha.sandstorm.io/grain/kB7bofATTL2jGuKSFwmqdA/GithubPRCommands for a list of commands.
